develop/devguide/GQI Extensions/Best Practices/GQI_Extensions_Caching.md

```diff
@@ -0,0 +1,104 @@
+- improves performance and reduces memory consumption by sharing results within a group.
+
+### Memory consumption
+
+While caching can greatly improve performance, it always comes at the cost of additional memory usage. The total memory footprint grows with both the size of each cached data set and the number of distinct cache entries being stored.
+
+A common example is caching per security group rather than per individual user. This approach is far more efficient than user-level caching, but it can still consume significant memory if your environment contains many groups or if each group's data set is large. Even when a single global cache is used (not per group), similar problems can occur if the data set itself is sizable or if stale entries are left in memory too long.
+
+To prevent excessive memory growth:
+
+- **Set limits** on the maximum number of groups cached at once.
+- **Implement eviction policies** (e.g. least recently used) to remove inactive groups.
+- **Use smaller data representations** (e.g. compress data or store only essential fields).
+- **Monitor memory usage** and adjust cache size or timeout values accordingly.
+
+### Cache invalidation
+
+Cache invalidation ensures that users do not see outdated or incorrect data. In ad hoc data sources, stale data can occur when the underlying data changes frequently or when security group memberships are updated. Without proper invalidation, cached results may remain inaccurate indefinitely.
+
+There are several strategies you can use:
+
+- **Time-based expiration**: Automatically refresh data after a fixed interval (e.g. 30 seconds). This is relatively simple to implement and works well when data updates are predictable.
+- **Manual reset (cache breaker)**: Use a file, flag, or external signal to clear or refresh the cache on demand. This is useful for administrative control when data changes are event-driven.
+- **Hybrid approach**: Combine timeouts with manual reset to balance predictable refresh with flexibility.
+- **Invalidate per security group**: If only certain groups are affected, clear cache entries selectively rather than globally.
+
+Carefully plan invalidation frequency. Excessively frequent invalidation negates the benefits of caching, while infrequent invalidation risks serving stale or incorrect data.
+
+### Concurrency
+
+In multi-user or multi-threaded environments, cache methods may be called concurrently by different threads. Without proper synchronization, this can cause race conditions, redundant data fetching, or an inconsistent cache state.
+
+To handle concurrency safely:
+
+- Use thread-safe collections such as `ConcurrentDictionary` for shared caches.
+- Protect critical sections (e.g. cache initialization or updates) with locks (`lock` statement or other synchronization primitives).
+- Double-check locking patterns help ensure initialization/update occurs only once per cache cycle, minimizing duplicate work.
```

Comment:
This sentence isn't entirely clear. Is this what you mean?
```suggestion
- Double-check whether locking patterns help ensure that initialization/updating occurs only once per cache cycle, minimizing duplicate work.
```

Reply:
["Double-checked locking"](https://en.wikipedia.org/wiki/Double-checked_locking) is a pattern.

So it should be something like:
```suggestion
- Use [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) to ensure that initializing/updating occurs only once per cache cycle, minimizing duplicate work.
```

Reply:
Not really, it's a concept that applies to programming in general. There is a paper and a book that explains it in detail (see e.g. https://www.dre.vanderbilt.edu/~schmidt/PDF/DC-Locking.pdf) and I found a section in https://learn.microsoft.com/en-us/archive/msdn-magazine/2005/october/understanding-low-lock-techniques-in-multithreaded-apps but I think that misses the point of explaining it to someone who's not familiar with it.